astropy/coordinates/solar_system.py

```diff
@@ -0,0 +1,302 @@
+
+
+def _get_earth_body_vector(time, body_key, earth_time=None):
+    """
+    Calculate the vector between the Geocenter and body with ``body_key``.
+
+    This routine calculates the vector between the Earth's Geocenter and the body
+    specified by ``body_key``.
+
+    Uses ``jplephem`` with the DE430 kernel.
+
+    Parameters
+    ----------
+    time : `~astropy.time.Time`
+        Time of observation.
+
+    body_key : int, str
+        The solar system body to calculate. Can be a string, e.g. 'moon',
+        or a valid integer index.
+
+        See help for ``get_body`` for a listing of valid integer indices.
+
+    earth_time : `~astropy.time.Time`
+        Time used for position of Earth. When correcting for light travel time,
+        one wants to use different times for the body in question and Earth.
+        If this is set to ```None```, the same time is used for both.
+
+    Returns
+    -------
+    earth_body_vector : `~astropy.coordinates.CartesianRepresentation`
+        Barycentric (ICRS) vector from Geocenter to the body in cartesian coordinates
+
+    earth_distance : `~astropy.units.Quantity`
+        Distance between Earth and body.
+
+    Notes
+    -----
+
+    """
+    earth_time = earth_time if earth_time is not None else time
+    earth_loc = get_barycentric_body_position(earth_time, 'earth')
+    body_loc = get_barycentric_body_position(time, body_key)
+
+    earth_body_vector = body_loc.xyz - earth_loc.xyz
+
+    earth_distance = np.sqrt(np.sum(earth_body_vector**2, axis=0))
+    return earth_body_vector, earth_distance
+
+
+def _get_apparent_body_position(time, body_key):
+    """
+    Calculate the apparent position of body ``body_key`` in cartesian
+    coordinates, given the approximate light travel time to the object.
+
+    Uses ``jplephem`` with the DE430 kernel.
+
+    Parameters
+    ----------
+    time : `~astropy.time.Time`
+        Time of observation
+
+    body_key : int, str
+        The solar system body to calculate. Can be a string, e.g. 'moon',
+        or a valid integer index.
+
+        See help for ``get_body`` for a listing of valid integer indices.
+
+    Returns
+    -------
+    cartesian_position : `~astropy.coordinates.CartesianRepresentation`
+        Barycentric (ICRS) apparent position of the body in cartesian coordinates
+    """
+    # Get distance of body at `time`
+    earth_to_body_vector, earth_distance = _get_earth_body_vector(time, body_key)
+
+    # The apparent position depends on the time that the light was emitted from
+    # the distant body, so subtract off the light travel time
+    light_travel_time = earth_distance/speed_of_light
+    emitted_time = time - light_travel_time
+
+    # Calculate position given approximate light travel time.
+    delta_light_travel_time = 20*u.s
+    while (np.fabs(delta_light_travel_time)) > 1.0e-8*u.s:
+        earth_to_body_vector, earth_distance = _get_earth_body_vector(emitted_time, body_key)
```

Comment:
To take into account that the earth position should be taken at a different time, you should pass in `time` as a third parameter, correct?

Also: I think the calculation before the `while` could just be done as the first iteration here.